### PR TITLE
VACMS-8077 Add Datadog APM to Local Environments and Tugboat

### PR DESCRIPTION
**Pull request by olivereri**
## Description
Modify development environment to send application tracing data to Datadog. Add this capability to Tugboat as well.

Relates to #8077 

## QA steps

As user _uid_ with _user_role_
1. Do this
2. Then that
3. Then validate Acceptance Criteria from issue
- [ ] This
- [ ] That
- [ ] The other thing

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Platform CMS Team`
- [ ] `Sitewide CMS Team`
  - [ ] `⭐️ Content ops`
  - [ ] `⭐️ CMS Experience`
  - [ ] `⭐️ Offices`
  - [ ] `⭐️ Product Support`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

File: .tugboat/config.yml [hunk 1]
```diff
@@ -41,6 +41,12 @@ services:
         - docker-php-ext-install opcache
         - a2enmod headers rewrite vhost_alias
 
+        # Install datadog agent for application tracing and profiling.
+        - DD_AGENT_MAJOR_VERSION=7 DD_API_KEY=e21aa8d73b3a48a8f82494bb91a096e7 DD_SITE="datadoghq.com" bash -c "$(curl -L https://s3.amazonaws.com/dd-agent/scripts/install_script.sh)"
+
+        # Install Datadog application traching PHP extension
```
Comment (ndouglas) [inline .tugboat/config.yml:47]:
```suggestion
        # Install Datadog application tracing PHP extension
```

File: .ddev/docker-compose.datadog.yaml.example [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      com.ddev.site-name: ${DDEV_SITENAME}
+    environment:
+      - DD_ENV=local-ddev
+      - DD_SERVICE=vagov-cms
+      - DD_API_KEY=
+      - DD_TAGS=user:yourUserName
+      - DD_LOGS_ENABLED=false
+      - DD_AGENT_HOST=datadog
+      - DD_TRACE_CLI_ENABLED=true
+      - DD_PROFILING_ENABLED=true
```
Comment (ndouglas) [inline .ddev/docker-compose.datadog.yaml.example:18]: Does DDEV support `~` for passing through values of environment variables if they are already set?

e.g. I was gonna do this for Lando:

Screen Shot 2022-03-09 at 8 23 35 AM

and then document `DD_API_KEY`, `DD_ENV`, etc in `.env.lando`.

I'm still figuring it out tho.

File: .ddev/.gitignore [hunk 3]
```diff
@@ -22,3 +22,5 @@
 /sequelpro.spf
 /xhprof
 /**/README.*
+!docker-compose.datadog.yaml.example
```
Comment (ndouglas) [inline .ddev/.gitignore:25]: Is this line needed?

**Comment by ndouglas** [comment]
`va/web/build` failed because of this:

```
Error: Entry Name "https://dev-va-gov-assets.s3-us-gov-west-1.amazonaws.com/generated/gi-sandbox.css" was not found.
    at Node.<anonymous> (/var/lib/tugboat/docroot/vendor/va-gov/content-build/src/site/stages/build/plugins/modify-dom/process-entry-names.js:147:15)
    at initialize.exports.each (/var/lib/tugboat/docroot/vendor/va-gov/content-build/node_modules/cheerio/lib/api/traversing.js:300:24)
    at Object.modifyFile (/var/lib/tugboat/docroot/vendor/va-gov/content-build/src/site/stages/build/plugins/modify-dom/process-entry-names.js:114:58)
    at Ware.<anonymous> (/var/lib/tugboat/docroot/vendor/va-gov/content-build/src/site/stages/build/plugins/modify-dom/index.js:49:18)
error Command failed with exit code 1.
```

Assuming that's a transient issue.

`/va/tests/behat` failed because of that prior `/va/web/build failure`, I believe.  And the `/va/tests/accessibility` failure is timeouts, so effectively random.

**TL;DR**: Test failures aren't related to PR content.

File: .ddev/docker-compose.datadog.yaml.example (hunk 2 above)
Comment (olivereri) [reply .ddev/docker-compose.datadog.yaml.example]: Didn't test if DDEV supported ~. However, instead of using .env.lando, I followed DDEV's pattern and broke the Datadog declaration and configuration to a .lando.local.yml.example file. Added a git ignore for .lando.local.yml hopefully making the setup for tracing easy.

File: .ddev/.gitignore (hunk 3 above)
Comment (olivereri) [reply .ddev/.gitignore]: Doesn't look like it. I removed it and git status don't show anything unneeded.

File: .ddev/docker-compose.datadog.yaml.example (hunk 2 above)
Comment (ndouglas) [reply .ddev/docker-compose.datadog.yaml.example]: fair 'nuff

**Review by ndouglas — DISMISSED** [review]

File: .ddev/web-build/Dockerfile [hunk 4]
```diff
@@ -2,3 +2,5 @@
 ARG BASE_IMAGE
 FROM $BASE_IMAGE
 ENV PATH=/var/www/html/vendor/bin:/var/www/html/bin:${PATH}
+#RUN sudo dpkg -i $(curl -w "%{filename_effective}" -LO $(curl -s https://api.github.com/repos/DataDog/dd-trace-php/releases | grep browser_download_url | grep 'amd64[.]deb' | head -n 1 | cut -d '"' -f 4))
+RUN sudo php $(curl -w "%{filename_effective}" -LO $(curl -s https://api.github.com/repos/DataDog/dd-trace-php/releases | grep browser_download_url | grep 'setup[.]php' | head -n 1 | cut -d '"' -f 4)) --enable-profiling --php-bin=php7.4
```
Comment (cweagans) [inline .ddev/web-build/Dockerfile:6]: Instead of specifying php7.4 here, I would recommend `$(basename $(realpath $(which php)))` or something along those lines. That way, we don't need to bump the PHP version here when we upgrade in the future.

File: .tugboat/datadog.yaml.j2 [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# Generated by Jinja2
+
+# Retrieve the Datadog API key and insert below. Should be enclosed in single quotes.
+api_key:
```
Comment (cweagans) [inline .tugboat/datadog.yaml.j2:4]: I haven't looked at any docs yet, but how does this work on the tugboat env? Would I have to commit the key to a PR branch?

File: .ddev/web-build/Dockerfile (hunk 4 above)
Comment (cweagans) [reply .ddev/web-build/Dockerfile:6]: Also this isn't urgent and shouldn't block the PR.

File: .tugboat/datadog.yaml.j2 (hunk 5 above)
Comment (cweagans) [reply .tugboat/datadog.yaml.j2:4]: Same thing here -- not urgent. Merge if you're happy with it.

**Review by cweagans — DISMISSED** [review]

File: .tugboat/datadog.yaml.j2 (hunk 5 above)
Comment (cweagans) [reply .tugboat/datadog.yaml.j2:4]: Makes total sense. Down the road if we want to start doing automated performance testing, we may have to figure out how to automatically enable this on all tugboat environments (maybe some global environment var thing in Tugboat or something along those lines), but that's definitely not a today problem. Thanks for the clarification here!
